src/datepicker/datepicker-input.ts

```diff
@@ -293,26 +300,26 @@ export class NgbInputDatepicker implements OnChanges,
 
   setDisabledState(isDisabled: boolean): void { this.disabled = isDisabled; }
 
-  validate(c: AbstractControl): {[key: string]: any} {
+  validate(c: AbstractControl): ValidationErrors | null {
     const {value} = c;
 
-    if (value === null || value === undefined) {
-      return null;
-    }
+    if (value != null) {
+      const ngbDate = this._fromDateStruct(this._dateAdapter.fromModel(value));
 
-    const ngbDate = this._fromDateStruct(this._dateAdapter.fromModel(value));
+      if (!this._calendar.isValid(ngbDate)) {
+        return {'ngbDate': {invalid: value}};
+      }
 
-    if (!this._calendar.isValid(ngbDate)) {
-      return {'ngbDate': {invalid: value}};
-    }
+      if (this.minDate && ngbDate !.before(NgbDate.from(this.minDate))) {
+        return {'ngbDate': {minDate: {minDate: this.minDate, actual: value}}};
+      }
 
-    if (this.minDate && ngbDate.before(NgbDate.from(this.minDate))) {
-      return {'ngbDate': {minDate: {minDate: this.minDate, actual: value}}};
+      if (this.maxDate && ngbDate !.after(NgbDate.from(this.maxDate))) {
+        return {'ngbDate': {maxDate: {maxDate: this.maxDate, actual: value}}};
+      }
     }
 
-    if (this.maxDate && ngbDate.after(NgbDate.from(this.maxDate))) {
-      return {'ngbDate': {maxDate: {maxDate: this.maxDate, actual: value}}};
-    }
+    return null;
```

Comment:
Assuming that this._fromDateStruct returns a valid NgbDateStruct (isValid is used inside _fromDateStruct) or null, we can just check that ngbDate is not null (otherwise, it's invalid).

Extra validation is done here. Instead, I would write:

```
      const ngbDate = this._fromDateStruct(this._dateAdapter.fromModel(value));

      if (!ngbDate) {
        return {'ngbDate': {invalid: value}};
      }

      if (this.minDate && ngbDate.before(NgbDate.from(this.minDate))) {
        return {'ngbDate': {minDate: {minDate: this.minDate, actual: value}}};
      }

      if (this.maxDate && ngbDate.after(NgbDate.from(this.maxDate))) {
        return {'ngbDate': {maxDate: {maxDate: this.maxDate, actual: value}}};
      }
```

Reply:
but then you would get an exit way without any return, pretty sure TypeScript will yell 😄

Reply:
Yes you're right, we still need the `return null;`, I just posted the changed code inside the if ;)

Reply:
@fbasso good catch, fixed it